# Homework Assignment (Problem Set) 3: 

Note: Problem Set 3 directly focuses on Modules 5 and 6: Integer Problems, Nonlinear and Multiobjective Programming

### Question 1: 
A fertilizer company just developed a new compound made up of two raw materials that is superior to existing fertilizers on the market. When raw materials $x_1$ and $x_2$ are combined, a quantity $q$ of fertilizer results given by: $q = 6x_1 + 4x_2 - 0.25x_1^2 - 0.125x_2^2$. The company wants to take advantage of this opportunity and produce as much as possible of the new fertilizer. The company has 180 dollars to buy raw materials at a unit price of 8 dollars and 5 dollars per unit, respectively. Further, the distributor can provide up to 10 units of raw material 1 ($x_1$)

#### Part A:
Formulate as a constrained nonlinear program to maximize the amount of this new fertilizer. Clearly indicate the variables, objective function, and constraints. 

#### Part B: 
Solve the Program (provide exact values for all variables and the optimal objective function). 

#### Notes
Quantity produced formula: 
$$q = 6x_1 + 4x_2 - 0.25x_1^2 - 0.125x_2^2$$

Given: 
- Budget: 180 dollars
- Price of $x_1$: 8 dollars/unit
- Price of $x_2$: 5 dollars/unit
- Supply limit on $x_1$: $\le 10$ units 

#### Part A: Formulate as a constrined nonlinear program to maximize the amount of this new fertilizer. Clearly indicate the variables, objective function, and constraints
Decision Variables: 
| Variable | Meaning |
| --- | --- |
| $x_1$ | Units of raw material 1 purchased |
| $x_2$ | Units of raw material 2 purchased |

Objective Function (max): 
$$\text{Max } q = 6x_1 + 4x_2 - 0.25x_1^2 - 0.125x_2^2$$

Constraints: 
$$8x_1 + 5x_2 \le 180 \text{ (budget)}$$
$$x_1 \le 10 \text{ (supply limit on } x_1\text{)}$$
$$x_1, x_2 \ge 0 \text{ (non-negativity)}$$

#### Part B: Solve the Program (provide exact values for all variables and the optimal Objective function).

Diminishing marginal returns: because of the negative terms $ - 0.25x_1^2 - 0.125x_2^2$, the contribution of $x_1$ to $q$ eventually turns negative, meaning at some point, buying more of $x_1$ actually hurts output. 

$q_{x1}$:
$$q_{x1} = 6x_1 - 0.25x_2^2$$
$$\frac{\partial q}{\partial x_1} = 6 - 0.5x_1$$
$$6 - 0.5x_1 = 0$$
$$0.5x_1 = 6$$
$$x_1 = 12$$

$q_{x2}$:
$$q_{x2} = 4x_2 - 0.125x_2^2$$
$$\frac{\partial q}{\partial x_2} = 4 - 0.25x_2$$
$$0.25x_2 = 4$$
$$x_2 = 16$$

$x_1 = 12$ violates the supply constraint so this can't be the optimal solution.

$x_2 = 16$ does not violate any constraints and $x_1$ and $x_2$ are separate in the objective function. Fixing $x_1 = 10$ has no effect on the optimal $x_2$. 

Optimal Solution: 
$$8(10) + 5(16) = 80 + 80 = 160 \le 180$$
$$x_1 = 10$$
$$x_2 = 16$$

Original Objective: 
$$q = 6x_1 + 4x_2 - 0.25x_1^2 - 0.125x_2^2$$
$$q = 6(10) + 4(16) - 0.25(10)^2 - 0.125(16)^2$$
$$q = 67$$

In [19]:
#pip install scipy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.3/20.3 MB 3.9 MB/s  0:00:05m0:00:0100:01

[notice] A new release of pip is available: 25.2 -> 26.1.1
[notice] To update, run: /Users/scottkeighley/.venv/bin/python -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [1]:
#pip install sympy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 6.7 MB/s  0:00:00 eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 536.2/536.2 kB 6.3 MB/s  0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [sympy]32m1/2 [sympy]

[notice] A new release of pip is available: 25.2 -> 26.1.1
[notice] To update, run: /Users/scottkeighley/.venv/bin/python -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [1]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import minimize
import sympy as sp

print("Libraries loaded successfully")

Libraries loaded successfully


In [9]:
# === Define the Objective Function ===
def q(x):
    x1, x2 = x
    return 6*x1 + 4*x2 - 0.25*x1**2 - 0.125*x2**2

def neg_q(x):
    return -q(x)

# Sanity check
test = [5, 10]
print(f"At x1={test[0]}, x2={test[1]}: q = {q(test):.3f}")
print(f"Budget check: 8({test[0]}) + 5({test[1]}) = ${8*test[0] + 5*test[1]} <= $180")

At x1=5, x2=10: q = 51.250
Budget check: 8(5) + 5(10) = $90 <= $180


In [11]:
# === Define Constraints and Bounds ===
constraints = [
    {'type': 'ineq', 'fun': lambda x: 180 - 8*x[0] - 5*x[1]},
    {'type': 'ineq', 'fun': lambda x: 10 - x[0]},
]
bounds = [(0, None), (0, None)]

print("Verifying test point [5, 10] is feasible:")
for i, c in enumerate(constraints):
    val = c['fun'](test)
    print(f" Constraint {i+1}: {val:.1f} >= 0? {'Yes' if val >=0 else 'No'}")

Verifying test point [5, 10] is feasible:
 Constraint 1: 90.0 >= 0? Yes
 Constraint 2: 5.0 >= 0? Yes


In [15]:
# === Run Optimizer ===
starting_points = [[1, 1], [5, 10], [10, 20], [8, 15], [9, 18]]

best_result = None
best_obj = -np.inf

print("Results from starting points:")
for x0 in starting_points: 
    result = minimize(neg_q, x0=x0, method='SLSQP',
                      bounds=bounds, constraints=constraints)

    obj_val = q(result.x)
    print(f" Start {x0}: x1={result.x[0]:.4f}, x2={result.x[1]:.4f}, q={obj_val:.4f}")
    if obj_val > best_obj:
        best_obj = obj_val
        best_result = result

x1_opt, x2_opt = best_result.x
print(f"\nBest: x1={x1_opt:.4f}, x2={x2_opt:.4f}, q={best_obj:.4f}")

Results from starting points:
 Start [1, 1]: x1=10.0000, x2=16.0000, q=67.0000
 Start [5, 10]: x1=10.0000, x2=16.0000, q=67.0000
 Start [10, 20]: x1=10.0000, x2=16.0000, q=67.0000
 Start [8, 15]: x1=10.0000, x2=16.0000, q=67.0000
 Start [9, 18]: x1=10.0000, x2=16.0000, q=67.0000

Best: x1=10.0000, x2=16.0000, q=67.0000


In [20]:
# === Verify Feasibility ===
x1_opt, x2_opt = best_result.x

budget_used = 8*x1_opt + 5*x2_opt
budget_slack = 180 - budget_used
supply_slack = 10 - x1_opt

print("=" * 50)
print("Verification")
print("=" * 50)
print(f" x1* = {x1_opt:.4f} units")
print(f" x2* = {x2_opt:.4f} units")
print(f" q* = {q(best_result.x):.4f} units of fertilizer")
print()
print("Constraint Status:")
budget_status = "Active" if abs(budget_slack) < 1e-4 else "inactive"
supply_status = "Active" if abs(supply_slack) < 1e-4 else "inactive"
print(f" Budget: ${budget_used:.2f} used of $180 (slack=${budget_slack:.2f}) {budget_status}")
print(f" Supply: x1={x1_opt:.4f} <= 10 (slack={supply_slack:.4f}) {supply_status}")
print(f"\n Optimizer status: {best_result.message}")

Verification
 x1* = 10.0000 units
 x2* = 16.0000 units
 q* = 67.0000 units of fertilizer

Constraint Status:
 Budget: $160.00 used of $180 (slack=$20.00) inactive
 Supply: x1=10.0000 <= 10 (slack=0.0000) Active

 Optimizer status: Optimization terminated successfully


### Question 2: 
South American River Company is looking to build a new fulfillment center to service deliveries to four towns (conveniently named 'Town A', 'Town B', and so on). Using cartesian coordinates, the locations of these towns and the number of shipments they receive each month are shown in the table below. All distances are in miles. 

| Town | Town Location (X) | Town Location (Y) | Shipments per month | 
| --- | --- | --- | --- |
| A | 5 | 10 | 150 |
| B | 10 | 5 | 200 |
| C | 6 | 2 | 175 |
| D | 0 | 12 | 125 |

#### Part A: 
Formulate the problem as a constrained nonlinear program that will enable us to minimize the total distance to satisfy these deliveries. Clearly indicate the variables, objective function, and constraints. Hint: how do we determine the distance between two points in the x-y plane?

#### Part B: 
Solve the Program (provide exact values for all variables and the optimal objective function). 

#### Part A: Formulate the problem as a constrained nonlinear program that will enable us to minimize the total distance to satisfy these deliveries. Clearly indicate the variables, objective function, and constraints. Hint: how do we determine the distance between two points in the x-y plane?

Decision Variables: 
| Variable | Meaning |
| --- | --- | 
| $x$ | X coordinate of the fulfillment center |
| $y$ | Y coordinate of the fulfillment center |

Distance Formulas: 

Distance A: 
$$d_A = \sqrt{(x-5)^2 + (y-10)^2}$$

Distance B: 
$$d_B = \sqrt{(x-10)^2 + (y-5)^2}$$

Distance C: 
$$d_C = \sqrt{(x-6)^2 + (y-2)^2}$$

Distance D: 
$$d_D = \sqrt{(x-0)^2 + (y-12)^2}$$

Objective Function: 
$$\text{Min } f(x,y) = 150d_A + 200d_B + 175d_C + 125d_D$$

Constraints:
$$x, y \ge 0$$

#### Part B: Solve the Program (provide exact values for all variables and the optimal objective function).

Try to get an estimated starting point by taking the average of each coordinate (?):
$$x_0 = \frac{5+10+6+0}{4} = \frac{21}{4} = 5.25$$
$$y_0 = \frac{10+5+2+12}{4} = \frac{29}{4} = 7.25$$

So, we could start roughly around (5,7)

In [40]:
towns = {
    'A': {'x': 5, 'y': 10, 'shipments': 150}, 
    'B': {'x': 10, 'y': 5, 'shipments': 200},
    'C': {'x': 6, 'y': 2, 'shipments': 175},
    'D': {'x': 0, 'y': 12, 'shipments': 125}
}

In [42]:
# === Define Objective Function ===
def d(coords):
    x, y = coords
    return (150 * np.sqrt((x-5)**2  + (y-10)**2) +
            200 * np.sqrt((x-10)**2 + (y-5)**2)  +
            175 * np.sqrt((x-6)**2  + (y-2)**2)  +
            125 * np.sqrt(x**2      + (y-12)**2))

# Sanity check at the simple average of town coordinates
x0 = (5 + 10 + 6 + 0) / 4   # = 5.25
y0 = (10 + 5 + 2 + 12) / 4  # = 7.25
print(f"Starting point: ({x0}, {y0})")
print(f"Objective at starting point: {d([x0, y0]):.4f}")

Starting point: (5.25, 7.25)
Objective at starting point: 3278.4561


In [44]:
# === Define Bounds & Run Optimizer ===
bounds = [(0, None), (0, None)]  # x >= 0, y >= 0

# Multiple starting points for robustness
starting_points = [
    [x0, y0],    
    [5, 5],
    [8, 8],
    [3, 7],
    [6, 6],
]

best_result = None
best_obj = np.inf

print("Results from multiple starting points:")
for x0_i in starting_points:
    result = minimize(d, x0=x0_i, method='SLSQP', bounds=bounds)
    obj_val = result.fun
    print(f"  Start {x0_i}: x={result.x[0]:.4f}, y={result.x[1]:.4f}, f={obj_val:.4f}")
    if obj_val < best_obj:
        best_obj = obj_val
        best_result = result

x_opt, y_opt = best_result.x
print(f"\nBest: x={x_opt:.4f}, y={y_opt:.4f}, f={best_obj:.4f}")

Results from multiple starting points:
  Start [5.25, 7.25]: x=6.8796, y=5.8060, f=3174.6994
  Start [5, 5]: x=6.8796, y=5.8060, f=3174.6994
  Start [8, 8]: x=6.8796, y=5.8060, f=3174.6994
  Start [3, 7]: x=6.8795, y=5.8059, f=3174.6994
  Start [6, 6]: x=6.8796, y=5.8059, f=3174.6994

Best: x=6.8796, y=5.8060, f=3174.6994


In [46]:
# === Verify Solution ===
x_opt, y_opt = best_result.x

# Individual distances
dA = np.sqrt((x_opt-5)**2  + (y_opt-10)**2)
dB = np.sqrt((x_opt-10)**2 + (y_opt-5)**2)
dC = np.sqrt((x_opt-6)**2  + (y_opt-2)**2)
dD = np.sqrt(x_opt**2      + (y_opt-12)**2)

print("=" * 55)
print("Solution Verification")
print("=" * 55)
print(f" Optimal location: x* = {x_opt:.4f}, y* = {y_opt:.4f}")
print()
print("Individual distances from optimal center:")
print(f" d_A = {dA:.4f} miles  (150 shipments: {150*dA:.2f} weighted)")
print(f" d_B = {dB:.4f} miles  (200 shipments: {200*dB:.2f} weighted)")
print(f" d_C = {dC:.4f} miles  (175 shipments: {175*dC:.2f} weighted)")
print(f" d_D = {dD:.4f} miles  (125 shipments: {125*dD:.2f} weighted)")
print()
print(f" Total weighted distance: f* = {best_obj:.4f}")
print()
print("Bound Status:")
print(f" x* = {x_opt:.4f} > 0  ← {'ACTIVE' if x_opt < 1e-4 else 'inactive (interior)'}")
print(f" y* = {y_opt:.4f} > 0  ← {'ACTIVE' if y_opt < 1e-4 else 'inactive (interior)'}")
print()
print(f" Optimizer status: {best_result.message}")

Solution Verification
 Optimal location: x* = 6.8796, y* = 5.8060

Individual distances from optimal center:
 d_A = 4.5960 miles  (150 shipments: 689.39 weighted)
 d_B = 3.2228 miles  (200 shipments: 644.56 weighted)
 d_C = 3.9063 miles  (175 shipments: 683.60 weighted)
 d_D = 9.2572 miles  (125 shipments: 1157.14 weighted)

 Total weighted distance: f* = 3174.6994

Bound Status:
 x* = 6.8796 > 0  ← inactive (interior)
 y* = 5.8060 > 0  ← inactive (interior)

 Optimizer status: Optimization terminated successfully


#### Solution:

Optimal Fulfillment Center Location: 
- x* = 6.8796 miles
- y* = 5.8060 miles

Minimum Total Weighted Distance: 
- f* = 3,174.6994 shipment miles

The optimal center at (6.88, 5.81) sits closest to Towns B and C, which are the two highest-volume towns. 

### Question 3: 
Local toy company, Toy-Vey, makes and distributes three different toys: tanks, trucks, and turtles. Tanks and Trucks each take two hours of labor while turtles only take one hour of labor. The manufacturing cost for one take is 7 dollars, for 1 truck is 5 dollars, and for 1 turtle is 4 dollars. Toy-Vey has a target budget of 164,000 dollars. Material requirements for the toys are shown below. 

| Toy | Plastic | Rubber | Metal |
| --- | --- | --- | --- |
| Tank | 2 | 1 | 2 |
| Truck | 3 | 1 | 1 |
| Turtle | 4 | 2 | 0 | 
| Available | 16,000 | 5,000 | 9,000 |

Management has ranked three goals it wishes to activate, arranged from highest to lowest priorities. 
- a) Minimize weekly overtime hours (over 10,000 hours) for production (40 hours for each of the 250 employees)
- b) Minimize over-utilization of the weekly available supply of materials used in making the toys and place twice as much emphasis on the plastic
- c) Minimize the under and over-utilization of the budget

Formulate the above decision problem as a single linear goal program. Do not solve
- Bonus (5 points): Solve the problem and give the number of each toy to produce as well as any violations of the goals (weights don't have to add up to 1; use simple weights - 1, 2, 3, etc.)

#### Notes

Decision Variables: 
| Variable | Meaning |
| --- | --- |
| $x_1$ | Number of tanks produced |
| $x_2$ | Number of trucks produced |
| $x_3$ | Number of turtles produced |

Goal programing - goals are soft, meaning we try to achieve them but allow deviations. To measure deviations from a goal, we need two variables for each goal: 
- $d^+$ = the amount we exceed the goal
- $d^-$ = the amount we fall short of the goal

(a) Because we want to minimize overtime hours (over 10,000 hours), we want to use $d^+$. The goal of (a) would be: 
$$\text{Minimize } d_1^+$$

Corresponding constraint: 
$$2x_1 + 2x_2 + x_3 + d_1^- - d_1^+ = 10,000$$

(b) Over-utilization means we're using more than what is available, so we've gone over the target. Because plastic gets twice as much emphasis, the contribution to the objective is: 
$$\text{Minimize } 2d_2^+ + d_3^+ + d_4^+$$

Corresponding constraints: 
$$2x_1 + 3x_2 + 4x_3 + d_2^- - d_2^+ = 16,000 \text{ (plastic)}$$
$$x_1 + x_2 + 2x_3 + d_3^- - d_3^+ = 5,000 \text{ (rubber)}$$
$$2x_1 + x_2 + 0x_3 + d_4^- - d_4^+ = 9,000 \text{ (metal)}$$

(c) We are penalized in both directions: 
$$\text{Minimize } d_5^- + d_5^+$$

Corresponding constraints: 
$$7x_1 + 5x_2 + 4x_3 + d_5^- - d_5^+ = 164,000 \text{ (budget)}$$

#### Formulate the above decision problem as a single linear goal program. Do not solve

Decision Variables: 
| Variable | Meaning |
| --- | --- |
| $x_1$ | Number of tanks produced |
| $x_2$ | Number of trucks produced |
| $x_3$ | Number of turtles produced |

Deviation Variables: 
| Variable | Meaning |
| --- | --- |
| $d_1^+, d_1^-$ | Over/under of labor hours vs 10,000 | 
| $d_2^+, d_2^-$ | Over/under of plastic vs 16,000 | 
| $d_3^+, d_3^-$ | Over/under of rubber vs 5,000 |
| $d_4^+, d_4^-$ | Over/under of metal vs 9,000 |
| $d_5^+, d_5^-$ | Over/under of budget vs 164,000 |

Objective:
$$\text{Minimize } d_1^+$$
$$\text{Minimize } 2d_2^+ + d_3^+ + d_4^+$$
$$\text{Minimize } d_5^- + d_5^+$$

Constraints: 
$$2x_1 + 2x_2 + x_3 + d_1^- - d_1^+ = 10,000$$
$$2x_1 + 3x_2 + 4x_3 + d_2^- - d_2^+ = 16,000 \text{ (plastic)}$$
$$x_1 + x_2 + 2x_3 + d_3^- - d_3^+ = 5,000 \text{ (rubber)}$$
$$2x_1 + x_2 + 0x_3 + d_4^- - d_4^+ = 9,000 \text{ (metal)}$$
$$7x_1 + 5x_2 + 4x_3 + d_5^- - d_5^+ = 164,000 \text{ (budget)}$$
$$x_1, x_2, x_3, d_1^+, d_1^-, d_2^+, d_2^-, d_3^+, d_3^-, d_4^+, d_4^-, d_5^+, d_5^- \ge 0 \text{ (nonnegativity)}$$

In [57]:
import numpy as np
from scipy.optimize import linprog
import matplotlib.pyplot as plt

n_vars = 13

print("Variable index reference:")
labels = ['x1', 'x2', 'x3', 'd1-', 'd1+', 'd2-', 'd2+', 'd3-', 'd3+', 'd4-', 'd4+', 'd5-', 'd5+']
for i, label in enumerate(labels):
    print(f" [{i}] {label}")

Variable index reference:
 [0] x1
 [1] x2
 [2] x3
 [3] d1-
 [4] d1+
 [5] d2-
 [6] d2+
 [7] d3-
 [8] d3+
 [9] d4-
 [10] d4+
 [11] d5-
 [12] d5+


In [59]:
# === Define the goal constraints ===
A_eq = [
    [2, 2, 1, 1, -1,  0,  0,  0,  0,  0,  0,  0,  0],  # labor
    [2, 3, 4, 0,  0,  1, -1,  0,  0,  0,  0,  0,  0],  # plastic
    [1, 1, 2, 0,  0,  0,  0,  1, -1,  0,  0,  0,  0],  # rubber
    [2, 1, 0, 0,  0,  0,  0,  0,  0,  1, -1,  0,  0],  # metal
    [7, 5, 4, 0,  0,  0,  0,  0,  0,  0,  0,  1, -1],  # budget
]
b_eq = [10000, 16000, 5000, 9000, 164000]

bounds = [(0, None)] * n_vars

print("Goal constraints defined")
print(f" A_eq shape: {np.array(A_eq).shape}")
print(f" b_eq: {b_eq}")

Goal constraints defined
 A_eq shape: (5, 13)
 b_eq: [10000, 16000, 5000, 9000, 164000]


In [63]:
# === Minimize overtime hours ===
c1 = [0]*n_vars
c1[4] = 1

res1 = linprog(c1, A_eq=A_eq, b_eq=b_eq, bounds=bounds, method='highs')

d1_plus_star = res1.fun
x1_s1 = res1.x

print("=" * 50)
print("Stage 1: Minimize overtime (d1+)")
print("=" * 50)
print(f" Optimal d1+* = {d1_plus_star:.4f} hours of overtime")
print(f" x1 (tanks) = {x1_s1[0]:.2f}")
print(f" x2 (trucks) = {x1_s1[1]:.2f}")
print(f" x3 (turtles) = {x1_s1[2]:.2f}")
print(f" Solver status: {res1.message}")

Stage 1: Minimize overtime (d1+)
 Optimal d1+* = 0.0000 hours of overtime
 x1 (tanks) = 0.00
 x2 (trucks) = 0.00
 x3 (turtles) = 0.00
 Solver status: Optimization terminated successfully. (HiGHS Status 7: Optimal)


In [67]:
# === Minimize material over utilization ===
A_ineq_s2 = [[0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]]  
b_ineq_s2 = [d1_plus_star]

c2 = [0]*n_vars
c2[6] = 2
c2[8] = 1
c2[10] = 1

res2 = linprog(c2, A_ub=A_ineq_s2, b_ub=b_ineq_s2,
               A_eq=A_eq, b_eq=b_eq, bounds=bounds, method='highs')

z2_star = res2.fun
x1_s2 = res2.x

print("=" * 50)
print("Stage 2: Minimize material over-utilization")
print("=" * 50)
print(f" Optimal Z2* = {z2_star:.4f}")
print(f" d2+ (plastic over) = {x1_s2[6]:.4f}")
print(f" d3+ (rubber over)  = {x1_s2[8]:.4f}")
print(f" d4+ (metal over)   = {x1_s2[10]:.4f}")
print(f" x1 (tanks)  = {x1_s2[0]:.2f}")
print(f" x2 (trucks) = {x1_s2[1]:.2f}")
print(f" x3 (turtles)= {x1_s2[2]:.2f}")
print(f" Solver status: {res2.message}")

Stage 2: Minimize material over-utilization
 Optimal Z2* = 0.0000
 d2+ (plastic over) = 0.0000
 d3+ (rubber over)  = 0.0000
 d4+ (metal over)   = 0.0000
 x1 (tanks)  = 0.00
 x2 (trucks) = 0.00
 x3 (turtles)= 0.00
 Solver status: Optimization terminated successfully. (HiGHS Status 7: Optimal)


In [69]:
# === Minimize budget deviation ===

A_ineq_s3 = [
    [0, 0, 0, 0, 1, 0, 0, 0, 0, 0,  0, 0, 0],  
    [0, 0, 0, 0, 0, 0, 2, 0, 1, 0,  1, 0, 0],  
]
b_ineq_s3 = [d1_plus_star, z2_star]

c3 = [0]*n_vars
c3[11] = 1  
c3[12] = 1  

res3 = linprog(c3, A_ub=A_ineq_s3, b_ub=b_ineq_s3,
               A_eq=A_eq, b_eq=b_eq, bounds=bounds, method='highs')

x_final = res3.x

print("=" * 50)
print("STAGE 3: Minimize Budget Deviation")
print("=" * 50)
print(f" d5- (budget under) = {x_final[11]:.4f}")
print(f" d5+ (budget over)  = {x_final[12]:.4f}")
print(f" x1 (tanks)   = {x_final[0]:.2f}")
print(f" x2 (trucks)  = {x_final[1]:.2f}")
print(f" x3 (turtles) = {x_final[2]:.2f}")
print(f" Solver status: {res3.message}")

STAGE 3: Minimize Budget Deviation
 d5- (budget under) = 131000.0000
 d5+ (budget over)  = 0.0000
 x1 (tanks)   = 4000.00
 x2 (trucks)  = 1000.00
 x3 (turtles) = -0.00
 Solver status: Optimization terminated successfully. (HiGHS Status 7: Optimal)


In [71]:
# === Final Answer ===
x1_f = round(x_final[0])
x2_f = round(x_final[1])
x3_f = round(x_final[2])

labor   = 2*x1_f + 2*x2_f + x3_f
plastic = 2*x1_f + 3*x2_f + 4*x3_f
rubber  = x1_f   + x2_f   + 2*x3_f
metal   = 2*x1_f + x2_f
budget  = 7*x1_f + 5*x2_f + 4*x3_f

print("=" * 57)
print("FINAL ANSWER SUMMARY")
print("=" * 57)
print()
print("Optimal Production Plan:")
print(f"  x1* = {x1_f:,} tanks")
print(f"  x2* = {x2_f:,} trucks")
print(f"  x3* = {x3_f:,} turtles")
print()
print("Goal Achievement:")
print(f"  P1 — Labor:   {labor:,} hrs vs 10,000 target"
      f" - {'OVERTIME: ' + str(labor-10000) + ' hrs' if labor > 10000 else 'No overtime'}")
print(f"  P2 — Plastic: {plastic:,} vs 16,000 available"
      f" - {'OVER by ' + str(plastic-16000) if plastic > 16000 else 'Within limit'}")
print(f"       Rubber:  {rubber:,} vs 5,000 available"
      f" - {'OVER by ' + str(rubber-5000) if rubber > 5000 else 'Within limit'}")
print(f"       Metal:   {metal:,} vs 9,000 available"
      f" - {'OVER by ' + str(metal-9000) if metal > 9000 else 'Within limit'}")
print(f"  P3 — Budget:  ${budget:,} vs $164,000 target"
      f" - {'Over by $' + str(budget-164000) if budget > 164000 else 'Under by $' + str(164000-budget) if budget < 164000 else 'Exact'}")

FINAL ANSWER SUMMARY

Optimal Production Plan:
  x1* = 4,000 tanks
  x2* = 1,000 trucks
  x3* = 0 turtles

Goal Achievement:
  P1 — Labor:   10,000 hrs vs 10,000 target - No overtime
  P2 — Plastic: 11,000 vs 16,000 available - Within limit
       Rubber:  5,000 vs 5,000 available - Within limit
       Metal:   9,000 vs 9,000 available - Within limit
  P3 — Budget:  $33,000 vs $164,000 target - Under by $131000


### Question 4: 
Breaking Ad is planning its advertising campaign for a customer's new product and is going to leverage podcasts and YouTube for its advertisements. The total number of exposures per 1,000 dollars is estimated to be 10,000 for podcasts and 7,500 for YouTube. The customer sees the campaign as successful if 750,000 people are reached and consider the campaign to be superbly successful if the exposures exceed 1 million people. The customer also wants to target its two largest age groups: 18-21 and 25-30. The total number of exposures per 1,000 dollars for these age groups are shown below. 

| Age Group | Podcasts | YouTube |
| --- | --- | --- |
| 18-21 | 2,500 | 3,000 | 
| 25-30 | 3,000 | 1,500 |
| Total Exposures | 10,000 | 7,500 |

Management has ranked five goals it wishes to achieve, arranged from highest to lowest priorities. 
- a) Successful campaign - at least 750,000 exposures
- b) Limit advertising costs to 100,000 dollars
- c) Limit podcast advertising costs to 70,000 dollars
- d) Superbly successful campaign - at least 1 million exposures
- e) Achieve at least 250,000 exposures for each of the two age groupos. However, as the 25-30 age group has more buying power, double the emphasis on this age group over the 18-21 age group

Formulate the above decision problem as a single linear goal program. Do not solve. 
- Bonus (5 points): Solve the problem and give the expenditures for each media advertising campaign as well as any violations of the goals (weights don't have to add up to 1; use simple weights - 1, 2, 3, etc.)

#### Formulate the above decision problem as a single linear goal program. 

Decision Variables: 
| Variable | Meaning |
| --- | --- | 
| $x_1$ | Dollars spent on podcast advertising |
| $x_2$ | Dollars spent on YouTube advertising |

Deviation Variables: 
| Variable | Meaning | 
| --- | --- | 
| $d_1^+, d_1^-$ | Over/under of total exposures vs 750K |
| $d_2^+, d_2^-$ | Over/under of total cost vs 100K (100 units) |
| $d_3^+, d_3^-$ | Over/under of podcast cost vs 70K (70 units) |
| $d_4^+, d_4^-$ | Over/under of total exposures vs 1,000,000 |
| $d_5^+, d_5^-$ | Over/under of 18-21 exposures vs 250,000 |
| $d_6^+, d_6^-$ | Over/under of 25-30 exposures vs 250,000 | 

Objectives (in order of priority):

$$P_1: \text{Minimize } d_1^- \text{ (at least 750K exposures)}$$
$$P_2: \text{Minimize } d_2^+ \text{ (total cost less than 100K)}$$
$$P_3: \text{Minimize } d_3^+ \text{ (podcast cost less than 70K)}$$
$$P_4: \text{Minimize } d_4^- \text{ (at least 1M exposures)}$$
$$P_5: \text{Minimize } d_5^- + 2d_6^- \text{ (age group exposures, 25-30 weighted double)}$$

Goal Constraints: 

Objective (a) is to minimize the amount we fall short of 750,000, so we minimize $d_1^-$:
$$P_1: \text{Minimize } d_1^-: 10,000x_1 + 7,500x_2 + d_1^- - d_1^+ = 750,000$$

Objective (b) is to minimize the amount we exceed the 100,000 total advertising cost (100K/1K = 100 units):
$$P_2: \text{Minimize } d_2^+: x_1 + x_2 + d_2^- - d_2^+ = 100$$

Objective (c) is to minimize the amount we exceed the 70,000 podcast advertising cost (70K/1K = 70 units):
$$P_3: \text{Minimize } d_3^+: x_1 + d_3^- - d_3^+ = 70$$

Objective (d) is to minimize the amount we fall short of 1,000,000: 
$$P_4: \text{Minimize } d_4^-: 10,000x_1 + 7,500x_2 + d_4^- - d_4^+ = 1,000,000$$

Objective (e) is to minimize the amount we fall short of 250,000 exposures, with 25-30 age group being twice as valuable: 
$$P_5: \text{Minimize } d_5^- + 2d_6^-: 2,500x_1 + 3,000x_2 + d_5^- - d_5^+ = 250,000 \text{ (18-21 age group)}$$
$$P_5: \text{Minimize } d_5^- + 2d_6^-: 3,000x_1 + 1,500x_2 + d_6^- - d_6^+ = 250,000 \text{ (25-30 age group)}$$

Nonnegativity: 
$$x_1, x_2, d_1^-, d_1^+, d_2^-, d_2^+, d_3^-, d_3^+, d_4^-, d_4^+, d_5^-, d_5^+, d_6^-, d_6^+ \ge 0$$

### Question 5: 
Drew Mann's Fruit Stand sells a variety of fruit and, during the harvest season, sells apples. Drew allocates 750 dollars for apples each season and he purchases the apples from three different orchards. Orchard 1 charges 5 dollars for a bushel of apples and Drew sells each bushel for 10 dollars. Orchard 2 charges 6 dollars per bushel and Drew sells each bushel for 11 dollars. Orchard 3 charges 8 dollars per bushel and Drew sells each bushel for 20 dollars. However, Orchard 3 is picky and, if Drew decides to purchase their apples, can only purchase between 20 and 40. So for Orchard 3, Drew can either purchase 0 or 20-40 bushels. Lastly, Orchard 1 only has 50 bushels available to sell. Drew is asking for assistance to determine how to maximize his revenue (don't consider the purchase price of apples)

#### Part A: 
Formulate this as a linear or nonlinear program by using an indicator variable to maximize the market's revenue

#### Part B: 
Solve the problem and give the solution (decision variables and objective function)

#### Part A: Formulate this as a linear or nonlinear program by using an indicator variable to maximize the market's revenue

Decision Variables: 
| Variable | Type | Meaning |
| --- | --- | --- |
| $x_1$ | Continuous | Bushels purchased from Orchard 1 |
| $x_2$ | Continuous | Bushels purchased from Orchard 2 | 
| $x_3$ | Continuous | Bushels purchased from Orchard 3 |
| $y$ | Binary {0,1} | 1 if Drew purchases from Orchard 3, 0 otherwise |

Objective: 
$$\text{Max Z} = 10x_1 + 11x_2 + 20x_3$$

Budget Constraint: 
$$5x_1 + 6x_2 + 8x_3 \le 750$$

Orchard 1 Constraint: 
$$x_1 \le 50$$

Orchard 3 Constraint: 
$$x_3 \ge 20y$$
$$x_3 \le 40y$$

Nonnegativity: 
$$x_1, x_2, x_3, y \ge 0$$

In [87]:
import pulp
import matplotlib.pyplot as plt
import numpy as np

In [91]:
# === Define the Problem and Decision Variables ===
prob = pulp.LpProblem("DrewManns_FruitStand", pulp.LpMaximize)

x1 = pulp.LpVariable('x1', lowBound=0, cat='Continuous') # orchard 1
x2 = pulp.LpVariable('x2', lowBound=0, cat='Continuous') # orchard 2
x3 = pulp.LpVariable('x3', lowBound=0, cat='Continuous') # orchard 3

y = pulp.LpVariable('y', cat='Binary') # 1 if buying from Orchard 3

print("Decision variables:")
print(f" x1: {x1} (Orchard 1 bushels, continuous)")
print(f" x2: {x2} (Orchard 2 bushels, continuous)")
print(f" x3: {x3} (Orchard 3 bushels, continuous)")
print(f" y: {y} (Orchard 3 indicator, binary)")

Decision variables:
 x1: x1 (Orchard 1 bushels, continuous)
 x2: x2 (Orchard 2 bushels, continuous)
 x3: x3 (Orchard 3 bushels, continuous)
 y: y (Orchard 3 indicator, binary)


In [93]:
# === Define Objective Function & Constraints ===
prob += 10*x1 + 11*x2 + 20*x3, "Maximize_Revenue"

prob += 5*x1 + 6*x2 + 8*x3 <= 750, "Budget"
prob += x1 <= 50, "Orchard1_Supply"
prob += x3 >= 20*y, "Orchard3_Min"
prob += x3 <= 40*y, "Orchard3_Max"

print("Problem formulation:")
print(prob)

Problem formulation:
DrewManns_FruitStand:
MAXIMIZE
10*x1 + 11*x2 + 20*x3 + 0
SUBJECT TO
Budget: 5 x1 + 6 x2 + 8 x3 <= 750

Orchard1_Supply: x1 <= 50

Orchard3_Min: x3 - 20 y >= 0

Orchard3_Max: x3 - 40 y <= 0

VARIABLES
x1 Continuous
x2 Continuous
x3 Continuous
0 <= y <= 1 Integer



In [95]:
# === Solve +++
prob.solve(pulp.PULP_CBC_CMD(msg=0))

print(f"Solver status: {pulp.LpStatus[prob.status]}")
print()
print("Raw solution values:")
print(f" x1 = {pulp.value(x1):.4f} bushels (Orchard 1)")
print(f" x2 = {pulp.value(x2):.4f} bushels (Orchard 2)")
print(f" x3 = {pulp.value(x3):.4f} bushels (Orchard 3)")
print(f" y = {pulp.value(y):.0f} (Orchard 3 indicator)")
print()
print(f" Objective (Revenue) = ${pulp.value(prob.objective):,.2f}")

Solver status: Optimal

Raw solution values:
 x1 = 50.0000 bushels (Orchard 1)
 x2 = 30.0000 bushels (Orchard 2)
 x3 = 40.0000 bushels (Orchard 3)
 y = 1 (Orchard 3 indicator)

 Objective (Revenue) = $1,630.00
